lizard_auth_server/views.py

```diff
@@ -304,3 +324,190 @@ def profile(self):
+
+        The user already is authenticated. Do not present a login form,
+        but give him what he wants right away. This can be used to
+        easily refresh expired tokens.
+
+        """
+        redirect_to = request.GET.get('next', '')
+        sso_key = request.GET.get('portal', '')
+
+        if redirect_to and not JWTView.is_url(redirect_to):
+            reason = _('Invalid `next` query string parameter.')
+            return HttpResponseBadRequest(reason, content_type='text/plain')
+
+        if not sso_key:
+            reason = _('Missing `portal` query string parameter.')
+            return HttpResponseBadRequest(reason, content_type='text/plain')
+
+        if not JWTView.is_portal(sso_key):
+            reason = _('Invalid `portal` query string parameter.')
+            return HttpResponseBadRequest(reason, content_type='text/plain')
+
+        self.redirect_to = redirect_to
+        self.sso_key = sso_key
+
+        if not (hasattr(request, 'user') and request.user.is_authenticated()):
+            # User has not been authenticated yet: return login form.
+            return super(JWTView, self).get(request, *args, **kwargs)
+
+        # User already is authenticated: return token,
+        # but not before verifying his portal access!
+
+        portal = Portal.objects.get(sso_key=sso_key)
+
+        if not UserProfile.has_profile(request.user, portal):
+            reason = _('You do not have access to this portal.')
+            return HttpResponseBadRequest(reason, content_type='text/plain')
+
+        self.token = self.get_token(request.user, portal)
+        return self._response(None)
+
+    def post(self, request, *args, **kwargs):
+        """Process login form."""
+        self.redirect_to = request.POST.get('next', '')
+        self.sso_key = request.POST.get('portal', '')
+        return super(JWTView, self).post(request, *args, **kwargs)
+
+    def get_context_data(self, **kwargs):
+        """Return a dictionary with template context."""
+        # The `next` hidden input element in the login template is not managed
+        # by the form class. For that reason, it's added to the context here.
+        context = super(JWTView, self).get_context_data(**kwargs)
+        context['next'] = self.redirect_to
+        return context
+
+    def get_initial(self):
+        """Return a dictionary for prepopulating the form."""
+        # Initialize the hidden `portal` field with the public key.
+        initial = super(JWTView, self).get_initial()
+        initial['portal'] = self.sso_key
+        return initial
+
+    def get_success_url(self):
+        """Return the `next` URL with token."""
+        # URL to parts.
+        scheme, netloc, path, params, query, fragment = parse.urlparse(
+            self.redirect_to)
+        # Add (extra) query string parameter.
+        data = parse.parse_qs(query)
+        data['access_token'] = self.token
+        query = parse.urlencode(data, True)
```

Comment:
Nee, alleen een reeds aanwezige access_token query string parameter. Dat lijkt me aanvaardbaar. Hebben we vanmiddag al mondeling besproken.